cpp/src/lists/contains.cu

```diff
@@ -105,48 +140,168 @@ struct search_list_fn {
+
+  search_list_nested_types_fn(duplicate_find_option const find_option,
+                              KeyValidityIter const& key_validity_iter,
+                              EqComparator const& d_comp,
+                              bool search_key_is_scalar)
+    : find_option(find_option),
+      key_validity_iter(key_validity_iter),
+      d_comp(d_comp),
+      search_key_is_scalar(search_key_is_scalar)
+  {
+  }
+
+  __device__ size_type operator()(list_device_view list) const
+  {
+    // A null list or null key will result in a null output row.
+    if (list.is_null() || !key_validity_iter[list.row_index()]) { return NULL_SENTINEL; }
+
+    return find_option == duplicate_find_option::FIND_FIRST ? search_list<true>(list)
+                                                            : search_list<false>(list);
+  }
 
-  /**
-   * @brief Return a pair of index iterators {begin, end} to loop through elements within a list.
-   *
-   * Depending on the value of `forward`, a pair of forward or reverse iterators will be
-   * returned, allowing to loop through elements in the list in first-to-last or last-to-first
-   * order.
-   *
-   * Note that the element indices always restart to `0` at the first position in each list.
-   *
-   * @tparam forward A boolean value indicating whether we want to iterate elements in the list by
-   *         forward or reverse order.
-   * @param size The number of elements in the list.
-   * @return A pair of {begin, end} iterators to iterate through the range `[0, size)`.
-   */
+ private:
   template <bool forward>
-  static __device__ auto element_index_pair_iter(size_type const size)
+  __device__ inline size_type search_list(list_device_view list) const
   {
-    if constexpr (forward) {
-      return thrust::pair(thrust::make_counting_iterator(0), thrust::make_counting_iterator(size));
+    using cudf::experimental::row::lhs_index_type;
+    using cudf::experimental::row::rhs_index_type;
+
+    auto const [begin, end] = element_index_pair_iter<forward>(list.size());
+    auto const found_iter =
+      thrust::find_if(thrust::seq, begin, end, [&] __device__(auto const idx) {
+        return !list.is_null(idx) &&
+               d_comp(static_cast<lhs_index_type>(list.element_offset(idx)),
+                      static_cast<rhs_index_type>(search_key_is_scalar ? 0 : list.row_index()));
+      });
+    // If the key is found, return its found position in the list from `found_iter`.
+    return found_iter == end ? NOT_FOUND_SENTINEL : *found_iter;
+  }
+};
+
+/**
+ * @brief Function to search for key element(s) in the corresponding rows of a lists column,
+ * specialized for non-nested types.
+ */
+template <bool search_key_is_scalar,
+          typename Element,
+          typename InputIterator,
+          typename SearchKeyType>
+void index_of_non_nested(InputIterator input_it,
+                         size_type num_rows,
+                         size_type* output_it,
```

Comment:
For symmetry (and generic iterators), I'd prefer to see a templated `OutputIterator` instead of `size_type*`.

Reply:
...and similarly below in `index_of_nested_types`.